operations/helm/charts/mimir-distributed/templates/_helpers.tpl

```diff
@@ -303,3 +342,38 @@ Cluster name that shows up in dashboard metrics
     {{- print "policy/v1beta1" -}}
   {{- end -}}
 {{- end -}}
+
+{{/* Creates dict for zone aware replication configuration */}}
+{{- define "mimir.zoneAwareReplicationMap" -}}
```

Comment:
There are two structures here of importance:
.Values.zone_aware_replication
.Values.<component>.zone_aware_replication.  

This merges the two with component overriding non.  There are many facets to what a person may want to configure for zone awareness.  In our case a zone is an aws AZ.  I believe for grafana its a virtual labelling and grouping of nodes.  So for us we want to ensure the pods are scheduled (via antiaffinity) into the appropriate AZs (topology.kubernetes.io/zone).  For grafana it would be something entirely different.  There's also the affinity one may desire to prevent pods from binpacking into the same hostname.  And then there's node selection.   We dont use it but others may have dedicated nodes for their mimir stuff entirely, or perhaps they just want the ingesters on different nodes.  We're close to that point in our prod env as our mimir ingesters are using nearly the entire RAM that can be provided by our default node groups RAM.  So we nearly need to create a larger instance type just for the ingesters.

The .Values.zone_aware_replication is for those that simply want to set a common nodeSelection/affinity across all the components with the ability to overwrite that definition via the .Values.component.zone...

If we want to get rid of the .Values.zone_aware_replication entirely and force users to individually define each .Values.component.z_a_r , I'm fine with that, it would definitely simply the code, but at the expense of some users ability to configure more easily.

Unfortunately due to the nature of this feature and what is a zone, I dont think we're going to make everyone happy.

Reply:
There is also the aspect of the default values.yaml that the chart provides along with any values files that a user supplies.    In order for values that a user supplies to overwrite the "default" configuration from the helm chart, the "map" has to be a list not a dict as helm merges dicts, but overwrites lists when dealing with combining values files and/or setting variables directly.

Reply:
Thanks for the thorough explanation! I agree we can't make everyone happy. I think we should prioritize two things:

1. Ensure the output for a given values file is relatively predictable. Nothing more annoying than running into template errors again and again when you already know exactly what you want to have in the final output.
2. Ensure a reasonable amount of flexibility to support 90% of use-cases. We can't get to 100% without adding a field for everything in kubernetes which defeats the purpose of a helm chart.

With that said, I think I'd prefer to eliminate the top-level `.Values.zone_aware_replication`. In most cases we will need to at least specify a different value for the `app.kubernetes.io/component` label. Either `store-gateway`, `ingester`, `alertmanager`, so I think it's probably relatively rare that a common zone configuration is possible. Same argument applies for node selectors, since these components have such different resource requirements it's unlikely users would want to move just those three onto a special node type I think.

Then in the (I assume rare) case that is needed, it's still possible, albeit with a bit of duplication.

WDYT?